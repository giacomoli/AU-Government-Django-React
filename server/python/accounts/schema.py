from datetime import datetime, timedelta

import graphene
from sitename.decorators import login_required, login_required_relay
from billing.filters import MatterFilter
from billing.models import Matter, Note, TimeEntry
from billing.schema import MatterType, NoteType
from core.connection import Connection
from core.mixins import LoginRequiredRelayMixin
from core.scalars import Decimal as DecimalScalar
from django.conf import settings
from django.db.models import Sum, Value
from django.db.models.functions import Coalesce
from graphene import relay
from graphene_django.filter import DjangoFilterConnectionField
from graphene_django.types import DjangoObjectType

from .filters import ClientFilter, ContactFilter, UserFilter
from .models import Client, Contact, Location, Organisation, User


class LocationType(DjangoObjectType):
    state = graphene.Int()
    state_display = graphene.String()
    country = graphene.String()

    def resolve_state(self, info):
        return self.state

    def resolve_state_display(self, info):
        return self.get_state_display()

    def resolve_country(self, info):
        if self.country == '0':
            return ''
        else:
            return self.country

    class Meta:
        model = Location


class PostalLocationType(graphene.ObjectType):
    id = graphene.ID()
    postal_address1 = graphene.String()
    postal_address2 = graphene.String()
    postal_suburb = graphene.String()
    postal_state = graphene.Int()
    postal_post_code = graphene.String()
    postal_country = graphene.String()


class LocationInterface(graphene.Interface):
    addresses_are_equals = graphene.Boolean()
    postal_location = graphene.Field(PostalLocationType)

    def resolve_addresses_are_equals(self, info):
        return self.location == self.postal_location

    def resolve_postal_location(self, info):
        if self.postal_location:
            return PostalLocationType(
                id=self.postal_location.id,
                postal_address1=self.postal_location.address1,
                postal_address2=self.postal_location.address2,
                postal_suburb=self.postal_location.suburb,
                postal_state=self.postal_location.state,
                postal_post_code=self.postal_location.post_code,
                postal_country=self.postal_location.country,

            )

        return None


class OrganisationType(LoginRequiredRelayMixin, DjangoObjectType):
    group_status = graphene.Int()
    link_mails = graphene.Boolean()

    def resolve_group_status(self, info):
        return self.group_status
    
    def resolve_link_mails(self, info):
        return self.link_mails

    class Meta:
        model = Organisation
        interfaces = (relay.Node, LocationInterface)
        connection_class = Connection
        filter_fields = {
            'name': ['icontains']
        }


class UserType(LoginRequiredRelayMixin, DjangoObjectType):
    full_name = graphene.String()
    photo = graphene.String()
    salutation = graphene.Int()
    access = graphene.Int()
    location = graphene.Field(lambda: LocationType)
    rate = DecimalScalar()
    groups = graphene.List(graphene.String)
    has_gmail_account = graphene.Boolean()
    pointer = graphene.Int()
    addresses_are_equals = graphene.Boolean()
    can_use_xero = graphene.Boolean()
    can_link_mails = graphene.Boolean()
    can_delete_mails = graphene.Boolean()
    gmail = graphene.String()
    units_today = graphene.Int()
    units_week = graphene.Int()
    units_month = graphene.Int()
    mail_enabled = graphene.Boolean()

    class Meta:
        model = User
        interfaces = (relay.Node,)
        connection_class = Connection

    def resolve_full_name(self, info):
        return self.full_name

    def resolve_photo(self, info):
        if self.photo.name:
            return str(self.photo.url)

    def resolve_pointer(self, info):
        user = self

        matters = list(
            user.manager_of_matters.filter(
                billable_status=1).values_list('matter_status', flat=True)
        ) + list(user.assistant_of_matters.filter(
            billable_status=1).values_list('matter_status', flat=True))

        all_open_matters = (list(
            user.manager_of_matters.values_list('billable_status', flat=True))
            + list(user.assistant_of_matters.values_list(
                'billable_status', flat=True))).count(1)
        if all_open_matters == 0:
            all_open_matters = 1
        active_high = matters.count(1)
        active_moderate = matters.count(2)
        active_low = matters.count(3)
        waiting_for_internal_review = matters.count(4)
        waiting_for_AA_review = matters.count(5)
        waiting_for_external = matters.count(6)
        ad_hoc_work = matters.count(7)
        need_to_be_build = matters.count(8)
        pointer = ((active_high * 10
                   + active_moderate * 7
                   + active_low * 5
                   + waiting_for_internal_review * 2
                   + waiting_for_AA_review * 2
                   + waiting_for_external * 1.5
                   + ad_hoc_work * 4
                   + need_to_be_build * 1) / all_open_matters) * 2
        return pointer

    def resolve_salutation(self, info):
        return self.salutation

    def resolve_access(self, info):
        return self.access

    def resolve_location(self, info):
        # If no location -> return empty Location instance with fake id
        return self.location or Location(id=-1)

    def resolve_rate(self, info):
        return self.rate

    def resolve_groups(self, info):
        return [group['name'] for group in self.groups.values('name')]

    def resolve_has_gmail_account(self, info):
        return hasattr(self, 'gmail_account')

    def resolve_addresses_are_equals(self, info):
        return self.location == self.postal_location

    def resolve_can_use_xero(self, info):
        return self.can_use_xero

    def resolve_can_link_mails(self, info):
        return self.can_link_mails

    def resolve_can_delete_mails(self, info):
        return self.can_delete_mails

    def resolve_gmail(self, info):
        return self.gmail

    def resolve_units_today(self, info):
        today = self.time_entries.filter(
            date__day=datetime.now().date().day,
            date__month=datetime.now().date().month,
            date__year=datetime.now().date().year
            ).filter(
            entry_type=1)
        units = today.aggregate(Sum('units'))['units__sum']
        print(today)
        print(units)
        return int(units) if units else 0

    def resolve_units_month(self, info):
        month = self.time_entries.filter(entry_type=1).filter(
            date__month=datetime.now().month,
            date__year=datetime.now().year
            )
        units = month.aggregate(Sum('units'))['units__sum']
        return int(units) if units else 0

    def resolve_units_week(self, info):
        date = datetime.today()
        start_week = date - timedelta(date.weekday()) - timedelta(1)
        end_week = start_week + timedelta(7)
        week = self.time_entries.filter(entry_type=1).filter(
            date__range=(start_week, end_week)
            )
        units = week.aggregate(Sum('units'))['units__sum']
        return int(units) if units else 0

    def resolve_mail_enabled(self, info):
        return self.mail_enabled


class ContactType(LoginRequiredRelayMixin, DjangoObjectType):
    full_name = graphene.String()
    photo = graphene.String()
    salutation = graphene.Int()
    occupation = graphene.Int()
    notes = graphene.List(NoteType)
    last_note = graphene.Field(NoteType)
    spouse = graphene.Field(lambda: ContactType)
    second_contact = graphene.Field(lambda: ContactType)
    children = graphene.List(lambda: ContactType)
    link_mails = graphene.Boolean()

    class Meta:
        model = Contact
        interfaces = (relay.Node, LocationInterface)
        connection_class = Connection

    def resolve_full_name(self, info):
        return self.full_name

    def resolve_photo(self, info):
        if self.photo.name:
            return "{}{}".format(settings.SITE_URL, self.photo.url)

        return None

    def resolve_salutation(self, info):
        return self.salutation

    def resolve_occupation(self, info):
        return self.occupation.id

    def resolve_spouse(self, info):
        return self.get_spouse()

    def resolve_children(self, info):
        return self.children

    def resolve_notes(self, info):
        return self.notes.all().order_by('-date_time')

    def resolve_last_note(self, info):
        return self.notes.last()

    def resolve_link_mails(self, info):
        return self.link_mails


class ClientType(LoginRequiredRelayMixin, DjangoObjectType):
    name = graphene.String()
    organisation = graphene.Field(OrganisationType)
    second_contact = graphene.Field(ContactType)
    invoicing_address = graphene.String()
    street_address = graphene.String()
    matters_count = graphene.Int()
    matters = DjangoFilterConnectionField(
        MatterType, exclude_status=graphene.Int(),
        filterset_class=MatterFilter)

    class Meta:
        model = Client
        interfaces = (relay.Node, )
        connection_class = Connection

    def resolve_name(self, info):
        return self.__str__()

    def resolve_organisation(self, info):
        if self.organisation:
            return self.organisation

        return Organisation(id=-1)

    def resolve_invoicing_address(self, info):
        if self.organisation:
            return self.organisation.formatted_postal_address

        return self.contact.formatted_postal_address

    def resolve_street_address(self, info):
        if self.organisation:
            return self.organisation.formatted_street_address

        return self.contact.formatted_street_address

    def resolve_matters_count(self, info):
        return self.matters.count()

    def resolve_second_contact(self, info):
        return self.second_contact

    def resolve_matters(self, info, exclude_status=None):
        if exclude_status:
            return self.matters.exclude(billable_status=exclude_status)

        return self.matters.all()


class Query:
    me = graphene.Field(UserType)
    user = relay.Node.Field(UserType)
    legal = DjangoFilterConnectionField(UserType, filterset_class=UserFilter)
    users = DjangoFilterConnectionField(UserType, filterset_class=UserFilter)
    contact = relay.Node.Field(ContactType)
    contacts = DjangoFilterConnectionField(
        ContactType,
        all=graphene.Boolean(),
        exclude=graphene.ID(),
        filterset_class=ContactFilter,
    )
    organisations = DjangoFilterConnectionField(OrganisationType)
    organisation = relay.Node.Field(OrganisationType)
    clients = DjangoFilterConnectionField(
        ClientType,
        filterset_class=ClientFilter
    )
    client = relay.Node.Field(ClientType)
    notes = graphene.List(NoteType, contact_id=graphene.ID())

    @login_required
    def resolve_users(self, info, **args):
        return User.objects.all()

    @login_required
    def resolve_clients(self, info, **args):
        return Client.objects.all()

    @login_required
    def resolve_me(self, info):
        return info.context.user

    @login_required_relay
    def resolve_contacts(self, info, exclude=None, all=False, **args):
        if all:
            if exclude:
                return Contact.objects.exclude(pk=exclude)
            return Contact.objects.all()
        else:
            if exclude:
                return Contact.objects.filter(is_general=False
                                              ).exclude(pk=exclude)

            return Contact.objects.filter(is_general=False)

        return Contact.objects.none()

    @login_required
    def resolve_organisations(self, info, **args):
        return Organisation.objects.all()

    @login_required
    def resolve_notes(self, info, contact_id=None):
        return Note.objects.filter(contact_id=contact_id).order_by('-date_time')

    @login_required
    def resolve_legal(self, info, **args):
        return User.objects.filter(is_legal=True)
