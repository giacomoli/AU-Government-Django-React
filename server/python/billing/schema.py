import graphene
from sitename.decorators import login_required, login_required_relay
from core.connection import Connection
from core.mixins import LoginRequiredRelayMixin
from core.scalars import Decimal as DecimalField
from django.db.models import Case, CharField, Q, Value, When
from django.db.models.functions import Concat
from graphene import relay
from graphene_django.filter import DjangoFilterConnectionField
from graphene_django.types import DjangoObjectType

from .filters import MatterFilter, TimeEntryFilter, InvoiceRecordsFilter
from .models import Matter, Note, StandartDisbursement, TimeEntry


class NoteType(DjangoObjectType):
    class Meta:
        model = Note


class TimeEntryType(LoginRequiredRelayMixin, DjangoObjectType):
    status_display = graphene.String()
    gst_status = graphene.Int()
    gst_status_display = graphene.String()
    cost = DecimalField()
    status = graphene.Int()
    billed_value = graphene.Float()
    rate = DecimalField()
    is_billed = graphene.Boolean()
    entry_type = graphene.Int()
    record_type = graphene.Int()

    class Meta:
        model = TimeEntry
        interfaces = (relay.Node,)
        connection_class = Connection

    def resolve_record_type(self, info):
        return self.time_entry_type.id

    def resolve_status_display(self, info):
        return self.get_status_display()

    def resolve_cost(self, info):
        return self.cost

    def resolve_status(self, info):
        return self.status

    def resolve_gst_status(self, info):
        return self.gst_status

    def resolve_gst_status_display(self, info):
        return self.get_gst_status_display()

    def resolve_billed_value(self, info):
        return self.billed_value

    def resolve_rate(self, info):
        if self.rate:
            return self.rate

        if self.staff_member:
            return self.staff_member.rate

        return 0

    def resolve_entry_type(self, info):
        return self.entry_type

    def resolve_is_billed(self, info):
        return bool(self.invoice)


class MatterType(LoginRequiredRelayMixin, DjangoObjectType):
    conflict_status = graphene.Int()
    billable_status = graphene.Int()
    billable_status_display = graphene.String()
    billing_method = graphene.Int()
    total_time_value = DecimalField()
    unbilled_time = graphene.List(TimeEntryType)
    total_time_invoiced = DecimalField()
    wip = DecimalField()
    amount_outstanding = DecimalField()
    is_paid = graphene.Boolean()
    days_open = graphene.Int()
    notes = graphene.List(NoteType)
    matter_status = graphene.Int()
    matter_status_display = graphene.String()
    last_note = graphene.Field(NoteType)
    total_invoiced_value = DecimalField()
    entry_type = graphene.Int()
    entry_type_display = graphene.String()
    matter_id = graphene.Int()
    lead_date = graphene.String()
    lead_status = graphene.Int()
    lead_status_display = graphene.String()

    class Meta:
        model = Matter
        interfaces = (relay.Node,)
        connection_class = Connection

    def resolve_lead_status_display(self, info):
        return self.lead_status.name

    def resolve_lead_status(self, info):
        return self.lead_status.id

    def resolve_lead_date(self, info):
        return self.lead_date

    def resolve_entry_type(self, info):
        return self.entry_type_id

    def resolve_entry_type_display(self, info):
        return self.entry_type.name

    def resolve_conflict_status(self, info):
        return self.conflict_status

    def resolve_billable_status(self, info):
        return self.billable_status

    def resolve_billing_method(self, info):
        return self.billing_method

    def resolve_billable_status_display(self, info):
        return self.get_billable_status_display()

    def resolve_total_time_value(self, info):
        return self.total_time_value

    def resolve_unbilled_time(self, info):
        return self.unbilled_time

    def resolve_total_time_invoiced(self, info):
        return self.total_time_invoiced(gst=False)

    def resolve_wip(self, info):
        return self.wip

    def resolve_amount_outstanding(self, info):
        return self.amount_outstanding

    def resolve_is_paid(self, info):
        return self.is_paid

    def resolve_days_open(self, info):
        return self.days_open

    def resolve_notes(self, info):
        return self.notes.all().order_by('-date_time')

    def resolve_matter_status(self, info):
        return self.matter_status_id

    def resolve_matter_status_display(self, info):
        return self.matter_status.name

    def resolve_last_note(self, info):
        return self.notes.last()

    def resolve_total_invoiced_value(self, info):
        return self.total_invoiced_value(gst=False)

    def resolve_matter_id(self, info):
        return self.id


class StandartDisbursementType(DjangoObjectType):
    gst_status = graphene.Int()

    class Meta:
        model = StandartDisbursement
        interfaces = (relay.Node,)
        filter_fields = {'name': ['icontains']}


class Query:
    matters = DjangoFilterConnectionField(
        MatterType, filterset_class=MatterFilter)
    matter = relay.Node.Field(MatterType)
    time_entries = DjangoFilterConnectionField(
        TimeEntryType, filterset_class=TimeEntryFilter)
    time_entry = relay.Node.Field(TimeEntryType)
    standart_disbursements = DjangoFilterConnectionField(
        StandartDisbursementType)
    notes = graphene.List(NoteType, matter_id=graphene.ID())
    invoice_records = DjangoFilterConnectionField(
        TimeEntryType, filterset_class=InvoiceRecordsFilter)

    @login_required
    def resolve_standart_disbursements(self, info, **args):
        return StandartDisbursement.objects.all()

    @login_required_relay
    def resolve_matters(self, info, **args):
        return Matter.objects.annotate(client_name=Case(When(
            Q(client__organisation__isnull=False) &
            Q(client__contact__isnull=False),
            then=Concat(
                'client__organisation__name',
                Value(' - '),
                'client__contact__first_name',
                Value(' '),
                'client__contact__last_name')
        ), When(
            Q(client__organisation__isnull=False) &
            Q(client__contact__isnull=True),
            then=Concat(
                'client__organisation__name',
                Value(' - '),
                Value('No Contact')
            )), When(
                Q(client__organisation__isnull=True) &
                Q(client__contact__isnull=False),
                then=Concat(
                    'client__contact__first_name',
                    Value(' '),
                    'client__contact__last_name'
                )), When(
                    Q(client__organisation__isnull=True) &
                    Q(client__contact__isnull=True),
                    then=Value(
                        'Have neither Organisation or Contact associated'
                    )),
            output_field=CharField())
        ).order_by('client_name')

    @login_required
    def resolve_time_entries(self, info, **args):
        return TimeEntry.objects.all()

    @login_required
    def resolve_notes(self, info, matter_id=None):
        return Note.objects.filter(matter_id=matter_id).order_by('-date_time')
