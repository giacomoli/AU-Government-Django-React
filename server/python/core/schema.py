import graphene
from sitename.decorators import login_required, login_required_relay
from core.connection import Connection
from django.db.models import FloatField
from django.db.models.functions import Cast
from graphene import relay
from graphene_django.filter import DjangoFilterConnectionField
from graphene_django.types import DjangoObjectType
from graphql_relay.node.node import from_global_id

from .models import Document
from .models import DocumentType as DocumentTypeModel
from .models import (Industry, InvoiceStatus, MatterSubType, MatterType,
                     Occupation, Office, Section)


class SectionType(DjangoObjectType):
    class Meta:
        model = Section


class DocumentTypeField(DjangoObjectType):
    class Meta:
        model = DocumentTypeModel


class DocumentType(DjangoObjectType):
    status = graphene.Int()
    status_display = graphene.String()
    nominated_type = graphene.Int()
    nominated_type_display = graphene.String()
    charging_clause = graphene.Int()
    charging_clause_display = graphene.String()

    class Meta:
        model = Document
        interfaces = (relay.Node,)
        filter_fields = []
        connection_class = Connection

    def resolve_status(self, info):
        return self.status

    def resolve_status_display(self, info):
        return self.get_status_display()

    def resolve_nominated_type(self, info):
        return self.nominated_type

    def resolve_nominated_type_display(self, info):
        return self.get_nominated_type_display()

    def resolve_charging_clause(self, info):
        return self.charging_clause

    def resolve_charging_clause_display(self, info):
        return self.get_charging_clause_display()


class IndustryType(DjangoObjectType):
    class Meta:
        model = Industry


class MatterTypeType(DjangoObjectType):
    class Meta:
        model = MatterType


class MatterSubTypeType(DjangoObjectType):
    class Meta:
        model = MatterSubType


class OccupationType(DjangoObjectType):
    class Meta:
        model = Occupation


class InvoiceStatusType(DjangoObjectType):
    class Meta:
        model = InvoiceStatus


class OfficeType(DjangoObjectType):
    suburb = graphene.String()
    name = graphene.String()
    short_name = graphene.String()

    def resolve_suburb(self, info):
        return self.location.suburb

    def resolve_name(self, info):
        return self.location.suburb

    def resolve_short_name(self, info):
        if self.location.suburb == 'Adelaide':
            return 'ADL'
        elif self.location.suburb == 'Sydney':
            return 'SYD'

        return 'Unknown'

    class Meta:
        model = Office


class Query:
    industries = graphene.List(IndustryType)
    matter_types = graphene.List(MatterTypeType)
    matter_sub_types = graphene.List(
        MatterSubTypeType, matter_type_id=graphene.ID()
    )
    invoice_statuses = graphene.List(InvoiceStatusType)
    offices = graphene.List(OfficeType)
    documents = DjangoFilterConnectionField(
        DocumentType,
        contact_id=graphene.ID(),
        organisation_id=graphene.ID()
    )
    document_types = graphene.List(DocumentTypeField)
    sections = graphene.List(SectionType)
    occupations = graphene.List(OccupationType)
    occupation = graphene.Field(
        OccupationType,
        occupation_id=graphene.ID()
    )

    @login_required
    def resolve_industries(self, info):
        return Industry.objects.all()

    @login_required
    def resolve_matter_types(self, info):
        return MatterType.objects.all()

    @login_required
    def resolve_matter_sub_types(self, info, matter_type_id=None):
        if matter_type_id:
            return MatterSubType.objects.filter(matter_type__id=matter_type_id)

        return MatterSubType.objects.all()

    @login_required
    def resolve_invoice_statuses(self, info):
        return InvoiceStatus.objects.all()

    @login_required
    def resolve_offices(self, info):
        return Office.objects.all()

    @login_required_relay
    def resolve_documents(self, info, **args):
        contact_id = args.get('contact_id')
        organisation_id = args.get('organisation_id')

        if contact_id:
            contact_id = from_global_id(contact_id)[1]
            return Document.objects.filter(contact__id=contact_id)

        if organisation_id:
            organisation_id = from_global_id(organisation_id)[1]
            return Document.objects.filter(organisation__id=organisation_id)

        return Document.objects.all()

    @login_required
    def resolve_document_types(self, info):
        return DocumentTypeModel.objects.all()

    @login_required
    def resolve_sections(self, info):
        return Section.objects.annotate(
                    section_number=Cast('number', FloatField())
                ).order_by('section_number').all()

    @login_required
    def resolve_occupations(self, info):
        return Occupation.objects.all()

    @login_required
    def resolve_occupation(self, info, occupation_id):
        if Occupation.objects.filter(id=occupation_id).exists():
            return Occupation.objects.get(id=occupation_id)
        return None
