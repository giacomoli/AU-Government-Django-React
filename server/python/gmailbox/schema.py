import datetime
from itertools import chain

import graphene
from sitename.decorators import login_required, login_required_relay
from core.connection import Connection
from core.mixins import LoginRequiredRelayMixin
from core.utils import get_paginator
from graphene import relay
from graphene_django.filter import DjangoFilterConnectionField
from graphene_django.types import DjangoObjectType
from graphql_relay.node.node import from_global_id
from billing.schema import MatterType

from .models import Mail, Attachment
from .filters import MailFilter


class AttachmentType(LoginRequiredRelayMixin, DjangoObjectType):
    class Meta:
        model = Attachment

class MailType(LoginRequiredRelayMixin, DjangoObjectType):
    sender_name = graphene.String()
    sender_address = graphene.String()
    recipient_name = graphene.String()
    recipient_address = graphene.String()
    subject = graphene.String()
    snippet = graphene.String()
    date = graphene.String()
    attachments = graphene.List(AttachmentType)
    matter = graphene.Field(MatterType)
    available_matters = graphene.List(MatterType)
    has_attachments = graphene.Boolean()

    class Meta:
        model = Mail
        interfaces = (relay.Node,)
        connection_class = Connection

    def resolve_sender_name(self, info):
        return self.sender_name

    def resolve_sender_address(self, info):
        return self.sender_address

    def resolve_recipient_name(self, info):
        return self.recipient_name

    def resolve_recipient_address(self, info):
        return self.recipient_address

    def resolve_subject(self, info):
        return self.subject

    def resolve_date(self, info):
        return self.date

    def resolve_attachments(self, info):
        return self.attachments.filter(inline=False)

    def resolve_matter(self, info):
        return self.matter

    def resolve_available_matters(self, info):
        return self.available_matters

    def resolve_has_attachments(self, info):
        return self.attachments.filter(inline=False).count() > 0


class Query:
    mails = DjangoFilterConnectionField(
        MailType, filterset_class=MailFilter
    )
    mail = relay.Node.Field(MailType)
    available_matters_of_mails = graphene.List(
        MailType,
        mails=graphene.List(graphene.ID)
    )

    @login_required_relay
    def resolve_mails(self, info, **args):
        return Mail.objects.filter(hidden=False)

    @login_required_relay
    def resolve_available_matters_of_mails(self, info, **args):
        mail_ids = [from_global_id(mail)[1] for mail in args.get('mails')]
        return Mail.objects.filter(id__in=mail_ids)
