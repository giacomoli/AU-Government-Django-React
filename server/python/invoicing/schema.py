from datetime import date, datetime
from itertools import chain

import graphene
from sitename.decorators import login_required_relay
from billing.filters import TimeEntryFilter
from billing.schema import TimeEntryType
from core.connection import Connection
from core.mixins import LoginRequiredRelayMixin
from core.scalars import Decimal
from graphene import relay
from graphene_django.converter import convert_django_field
from graphene_django.filter import DjangoFilterConnectionField
from graphene_django.types import DjangoObjectType
from simple_history.models import HistoricalRecords

from .filters import InvoiceFilter
from .models import Invoice, Payment
from .scalars import HistoricalRecord


@convert_django_field.register(HistoricalRecords)
def convert_location_field(field, registry=None):
    return HistoricalRecord()


class PaymentType(DjangoObjectType):
    method_display = graphene.String()
    amount = Decimal()

    def resolve_method_display(self, info):
        return self.get_method_display()

    def resolve_amount(self, info):
        return self.amount

    class Meta:
        model = Payment


class InvoiceType(LoginRequiredRelayMixin, DjangoObjectType):
    status_display = graphene.String()
    value_ex_GST = Decimal()
    value_incl_GST = Decimal()
    received_payments = Decimal()
    net_outstanding = Decimal()
    history = graphene.List(HistoricalRecord)
    total_billed_value = Decimal()
    is_paid = graphene.Boolean()
    due_date = graphene.String()
    number = graphene.ID()
    friendly_reminder = graphene.Boolean()
    first_reminder = graphene.Boolean()
    second_reminder = graphene.Boolean()
    time_entries = DjangoFilterConnectionField(
        TimeEntryType, filterset_class=TimeEntryFilter)
    billing_method = graphene.Int()
    time_entry_value = Decimal()
    fixed_price_value = Decimal()
    can_send_xero = graphene.Boolean()
    is_in_xero = graphene.Boolean()

    class Meta:
        model = Invoice
        interfaces = (relay.Node,)
        connection_class = Connection

    def resolve_time_entry_value(self, info):
        return self.time_entry_value

    def resolve_fixed_price_value(self, info):
        return self.fixed_price_value

    def resolve_status_display(self, info):
        return self.status.name

    def resolve_value_ex_GST(self, info):
        return self.value()

    def resolve_value_incl_GST(self, info):
        return self.value(gst=True)

    def resolve_received_payments(self, info):
        return self.received_payments

    def resolve_net_outstanding(self, info):
        return self.net_outstanding

    def resolve_history(self, info):
        return sorted(chain(
            self.history.exclude(history_change_reason__isnull=True),
            *[payment.history.exclude(history_change_reason__isnull=True)
              for payment in self.payments.all()]
        ), key=lambda x: x.history_date, reverse=True)

    def resolve_total_billed_value(self, info):
        return self.total_billed_value

    def resolve_is_paid(self, info):
        return self.is_paid

    def resolve_due_date(self, info):
        return self.due_date

    def resolve_number(self, info):
        return self.number

    def resolve_friendly_reminder(self, info):
        try:
            return not self.is_paid and \
                (date.today() - self.due_date).days >= 14
        except TypeError:
            return not self.is_paid and \
                (datetime.now() - self.due_date).days >= 14

    def resolve_first_reminder(self, info):
        try:
            return not self.is_paid and \
                (date.today() - self.due_date).days >= 28
        except TypeError:
            return not self.is_paid and \
                (datetime.now() - self.due_date).days >= 28

    def resolve_second_reminder(self, info):
        try:
            return not self.is_paid and \
                (date.today() - self.due_date).days >= 42
        except TypeError:
            return not self.is_paid and \
                (datetime.now() - self.due_date).days >= 42

    def resolve_time_entries(self, info, **args):
        return self.time_entries.all()

    def resolve_billing_method(self, info):
        return self.billing_method

    def resolve_can_send_xero(self, info):
        return self.can_send_xero

    def resolve_is_in_xero(self, info):
        return self.is_in_xero


class Query:
    invoices = DjangoFilterConnectionField(
        InvoiceType, filterset_class=InvoiceFilter)
    invoice = relay.Node.Field(InvoiceType)

    @login_required_relay
    def resolve_invoices(self, info, **args):
        return Invoice.objects.all()
