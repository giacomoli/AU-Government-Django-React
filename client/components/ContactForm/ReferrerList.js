import React, { Fragment } from 'react'

import AsyncAutocomplete from 'components/AsyncAutocomplete'

export default ({ referrers, ...props }) => (
  <Fragment>
    <div className="form-group">
      <AsyncAutocomplete link="contact" {...props} />
    </div>
    <div className="table-responsive">
      <table className="table table-bordered">
        <thead>
          <tr>
            <th>People referred by this Contact</th>
          </tr>
        </thead>
        <tbody>
          {referrers.length > 0 ? (
            referrers.map(o => (
              <tr key={o.node.id}>
                <td>
                  <a href={`/contact/${o.node.id}`} target="_blank">
                    {o.node.fullName}
                  </a>
                </td>
              </tr>
            ))
          ) : (
            <tr>
              <td>No referrers</td>
            </tr>
          )}
        </tbody>
      </table>
    </div>
  </Fragment>
)
