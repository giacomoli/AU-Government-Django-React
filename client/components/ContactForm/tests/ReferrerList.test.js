import React from 'react'
import renderer from 'react-test-renderer'
import { ApolloProvider } from 'react-apollo';
import { reducer } from 'redux-form'
import { ApolloClient } from 'apollo-client';
import { mockNetworkInterfaceWithSchema } from 'apollo-test-utils';
import { createStore } from 'redux';

import schema from './mocks/schema'
import ReferrerList from '../ReferrerList'

/* eslint-disable */
describe('Contact ReferrerList tab', () => {

  const mockNetworkInterface = mockNetworkInterfaceWithSchema({ schema });
  const client = new ApolloClient({
    networkInterface: mockNetworkInterface,
  })
  const store = createStore(reducer)

  it('should render ReferrerList tab without data', () => {

    const wrapper = renderer.create(
      <ApolloProvider client={client} store={store}>
        <ReferrerList
          label="This contact was introduced by"
          fieldName="referrer"
          onEdit={jest.fn()}
          onSelect={jest.fn()}
          getOptions={jest.fn()}
          value={{ name: 'text' }}
          accessor="fullName"
          placeholder="Search for contact"
          referrers
        />
      </ApolloProvider>
    )
    expect(wrapper).toMatchSnapshot()
  })

});
