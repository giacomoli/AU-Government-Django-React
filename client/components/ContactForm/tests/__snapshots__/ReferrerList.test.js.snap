// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact ReferrerList tab should render ReferrerList tab without data 1`] = `
Array [
  <div
    className="form-group"
  >
    <div>
      <label
        htmlFor="referrer"
      >
        This contact was introduced by
      </label>
      <div>
        <div
          className="Select is-clearable is-loading is-searchable Select--single"
        >
          <div
            className="Select-control"
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={null}
          >
            <span
              className="Select-multi-value-wrapper"
              id="react-select-2--value"
            >
              <div
                className="Select-placeholder"
              >
                Loading...
              </div>
              <div
                className="Select-input"
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-activedescendant="react-select-2--value"
                  aria-expanded="false"
                  aria-haspopup="false"
                  aria-owns=""
                  id="referrer"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  required={false}
                  role="combobox"
                  style={
                    Object {
                      "boxSizing": "content-box",
                      "width": "5px",
                    }
                  }
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </span>
            <span
              aria-hidden="true"
              className="Select-loading-zone"
            >
              <span
                className="Select-loading"
              />
            </span>
            <span
              className="Select-arrow-zone"
              onMouseDown={[Function]}
            >
              <span
                className="Select-arrow"
                onMouseDown={[Function]}
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="table-responsive"
  >
    <table
      className="table table-bordered"
    >
      <thead>
        <tr>
          <th>
            People referred by this Contact
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            No referrers
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
]
`;
